cirq/contrib/quirk/cells/composite_cell.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, TYPE_CHECKING, Callable, Optional, Iterator, cast, \
+    Iterable, TypeVar
+
+from cirq import circuits
+from cirq.contrib.quirk.cells.cell import Cell
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class CompositeCell(Cell):
+    """A cell made up of a grid of sub-cells.
+
+    This is used for custom circuit gates.
+    """
+
+    def __init__(self, height: int,
+                 sub_cell_cols_generator: Iterable[List[Optional[Cell]]], *,
+                 weight: int):
+        self.height = height
+        self._sub_cell_cols_generator = sub_cell_cols_generator
+        self._weight = weight
+        if weight <= 0:
+            self._sub_cell_cols_generator = []
+
+    def gate_count(self) -> int:
+        return self._weight
+
+    def _transform_cells(self, func: Callable[[Cell], Cell]) -> 'CompositeCell':
+        return CompositeCell(
+            height=self.height,
+            # It is important that this is a generator instead of a list!
+            # We must not list until after global operation counting.
```

Comment:
IIUC, you're implying that client code may modify the contents of the generator `sub_cell_cols_generator` after a composite cell is instantiated. Where does this happen? I only found to places of instantiation: one in `_parse_cols_into_composite_cell` which doesn't need this functionality since it passes in a list and one here in `_transform_cells` which may be reimplemented not to needed it.

What am I missing? Why do we need this?

Reply:
`self._sub_cell_cols_generator` may be exponentially large. We can define a generator over it, and lazily cache its results, but we must not actually run that generator yet because the top-level decision of "this circuit is too big?" has not yet been made.